### PR TITLE
Replace embedded libgc with Unity fork of recent Boehm (bdwgc)

### PR DESCRIPTION
**Pull request by joncham**
- Remove old embedded Boehm source
- Add bdwgc submodule
- Adjust build scripts as needed
- Use non-deprecated Boehm APIs in mono source
- Build boehm GC as single source file; this helps with performance
- Boehm submodule includes changes for building static libraries in tree for testing (nodist)

### CODE REVIEW

**Comment by vargaz** [comment]
RIP.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
Some lanes fail because of:
Making test-bundle in external/bdwgc
make[1]: *** No rule to make target 'test-bundle'.  Stop.

An empty rule needs to be added to libgc/Makefile.am:
test-bundle:

File: mono/metadata/boehm-gc.c [hunk 1]
```diff
@@ -127,6 +121,7 @@ mono_gc_base_init (void)
 	mono_w32handle_init ();
 #endif
 
+#if JON
```
Comment (CoffeeFlux) [inline mono/metadata/boehm-gc.c]: Left over from debugging?

Comment (joncham) [reply mono/metadata/boehm-gc.c]: Yes, this needs removed. All access in this block is using deprecated APIs and setting stackbottom is not needed.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by joncham** [comment]
Shall I squash, rebase, and fix a few of the commented areas?

**Comment by akoeplinger** [comment]
@lateralusX looks like the build still worked: https://dev.azure.com/dnceng/public/_build/results?buildId=435839&view=logs.

I can fix this up if you think it's important in my other PR so we can get this finally in :)

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lateralusX** [comment]
@akoeplinger OK lets fix that in your upcoming PR then.

**Comment by directhex** [comment]
Is it intended for `make install` in Mono to do a full install of Boehm?

```
[2019-11-23T11:29:35.458Z] DEBUG: BUILDSTDERR: error: Installed (but unpackaged) file(s) found:
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/cord.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/cord_pos.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/ec.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_allocator.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_backptr.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_config_macros.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_disclaim.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_gcj.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_inline.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_mark.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_pthread_redirects.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_tiny_fl.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_typed.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/gc_version.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/javaxfc.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/leak_detector.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/include/gc/weakpointer.h
06:29:35  DEBUG: BUILDSTDERR:    /usr/lib64/pkgconfig/bdw-gc.pc
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/AUTHORS
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.DGUX386
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.Mac
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.OS2
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.amiga
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.arm.cross
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.autoconf
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.cmake
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.cords
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.darwin
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.environment
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.ews4800
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.hp
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.linux
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.macros
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.rs6000
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.sgi
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.solaris2
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.symbian
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.uts
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.win32
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/README.win64
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/debugging.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/finalization.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/gcdescr.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/gcinterface.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/leak.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/overview.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/porting.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/scale.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/simple_example.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/doc/gc/tree.md
06:29:35  DEBUG: BUILDSTDERR:    /usr/share/man/man3/gc.3.gz
```

**Comment by lateralusX** [comment]
that doesn't look right, I assume we didn't do this for the old Boehm GC as part of make install, so should probably be excluded in updated version as well.

**Comment by rdp** [comment]
Is the unity fork better? Why replace?

**Comment by vargaz** [comment]
It is better, and mono doesn't really use it anymore, the main user is unity.

**Pull request merged**